lib/tapi_rdma/tapi_rdma.c

```diff
@@ -0,0 +1,253 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright (c) 2024 Advanced Micro Devices, Inc. */
+/** @file
+ * @brief Generic Test API to interact with RDMA links
+ *
+ * Generic API to interact with RDMA links.
+ */
+#define TE_LGR_USER     "TAPI RDMA"
+
+#include "te_config.h"
+#include "te_alloc.h"
+#include "te_string.h"
+#include "te_vector.h"
+#include "log_bufs.h"
+#include "tapi_job.h"
+#include "tapi_job_factory_rpc.h"
+#include "tapi_rdma.h"
+
+#define RDMA_TOOL "rdma"
+
+/* See description in tapi_rdma.h */
+te_errno
+tapi_rdma_link_get_stats(rcf_rpc_server *rpcs, const char *link,
+                         tapi_rdma_link_stats_t **stats)
+{
+    te_errno            rc;
+    tapi_job_factory_t *factory = NULL;
+    te_vec              stat_vec = TE_VEC_INIT(tapi_rdma_link_stat_t);
+    tapi_job_t         *job = NULL;
+    tapi_job_channel_t *chan_out;
+    tapi_job_channel_t *chan_err;
+    tapi_job_channel_t *filter_out;
+    tapi_job_buffer_t   buf_out = TAPI_JOB_BUFFER_INIT;
+    tapi_job_status_t   status;
+
+    int   ret;
+    char *output;
+    char *link_name = NULL;
+    int   consumed;
+
+    const char *argv[] = {
+        RDMA_TOOL,
+        "statistic",
+        "show",
+        "link",
+        link,
+        NULL,
+    };
+
+    tapi_job_simple_desc_t job_desc = {
+        .program = RDMA_TOOL,
+        .argv = argv,
+        .stdout_loc = &chan_out,
+        .stderr_loc = &chan_err,
+        .job_loc = &job,
+        .filters = TAPI_JOB_SIMPLE_FILTERS(
+            {
+                .use_stdout = true,
+                .filter_name = "Stdout",
+                .readable = true,
+                .filter_var = &filter_out,
+            },
+            {
+                .use_stderr = true,
+                .filter_name = "Stderror",
+                .log_level = TE_LL_ERROR,
+            }
+        ),
+    };
+
+    rc = tapi_job_factory_rpc_create(rpcs, &factory);
+    if (rc != 0)
+        goto out;
+
+    rc = tapi_job_simple_create(factory, &job_desc);
+    if (rc != 0)
+        goto out;
+
+    rc = tapi_job_start(job);
+    if (rc != 0)
+        goto out;
+
+    do {
+        rc = tapi_job_receive(TAPI_JOB_CHANNEL_SET(filter_out), -1, &buf_out);
+
+        if (rc != 0)
+            goto out;
+    } while (!buf_out.eos);
+
+    rc = tapi_job_wait(job, -1, &status);
+    if (rc != 0)
+        goto out;
+
+    if (status.type != TAPI_JOB_STATUS_EXITED || status.value != 0)
+    {
+        ERROR("RDMA utility finished abnormally");
+        rc = TE_RC(TE_TAPI, TE_EFAULT);
```

Comment:
1. `TE_EFAULT` does not mean "generic failure", it means "invalid address", and while it is used throughout TE to mean the former, it's not right.  The generic error code is `TE_EFAIL`.
2. That said, if a function returns the same error code for totally unrelated conditions, that's not good.